kube/src/api/subresource.rs

```diff
@@ -208,3 +211,177 @@ where
+    /// Defaults to only container if there is only one container in the pod.
+    pub container: Option<String>,
+    /// If `true`, the contents will be redirected to the standard input stream of the pod. Defaults to `false`.
+    pub stdin: bool,
+    /// If `true`, the standard out stream of the pod will be redirected to it. Defaults to `true`.
+    pub stdout: bool,
+    /// If `true`, the standard error stream of the pod will be redirected to it. Defaults to `true`.
+    pub stderr: bool,
+    /// If `true`, TTY will be allocated for the attach call. Defaults to `false`.
+    pub tty: bool,
+}
+
+#[cfg(feature = "ws")]
+impl Default for AttachParams {
+    // Default matching the server's defaults.
+    fn default() -> Self {
+        Self {
+            container: None,
+            stdin: false,
+            stdout: true,
+            stderr: true,
+            tty: false,
+        }
+    }
+}
+
+#[cfg(feature = "ws")]
+impl Resource {
+    /// Attach to a pod
+    pub fn attach(&self, name: &str, ap: &AttachParams) -> Result<http::Request<()>> {
+        let base_url = self.make_url() + "/" + name + "/" + "attach?";
+        let mut qp = url::form_urlencoded::Serializer::new(base_url);
+
+        if ap.stdin {
+            qp.append_pair("stdin", "true");
+        }
+        if ap.stdout {
+            qp.append_pair("stdout", "true");
+        }
+        if ap.stderr {
+            qp.append_pair("stderr", "true");
+        }
+        if ap.tty {
+            qp.append_pair("tty", "true");
+        }
+        if let Some(container) = &ap.container {
+            qp.append_pair("container", &container);
+        }
+
+        let req = http::Request::get(qp.finish());
+        req.body(()).map_err(Error::HttpError)
+    }
+}
+
+/// Marker trait for objects that has attach
+#[cfg(feature = "ws")]
+pub trait AttachingObject {}
+
+#[cfg(feature = "ws")]
+impl AttachingObject for k8s_openapi::api::core::v1::Pod {}
+
+#[cfg(feature = "ws")]
+impl<K> Api<K>
+where
+    K: Clone + DeserializeOwned + AttachingObject,
+{
+    /// Attach to pod
+    pub async fn attach(&self, name: &str, ap: &AttachParams) -> Result<AttachedProcess> {
+        let req = self.resource.attach(name, ap)?;
+        let stream = self.client.connect(req).await?;
+        Ok(AttachedProcess::new(stream, ap.stdin, ap.stdout, ap.stderr))
+    }
+}
+
+// ----------------------------------------------------------------------------
+// Exec subresource
+// ----------------------------------------------------------------------------
+/// Params for execution
+///
+/// Note that the server rejects when none of `stdin`, `stdout`, `stderr` are attached.
+#[cfg(feature = "ws")]
+pub struct ExecParams {
+    /// The container in which to execute the command.
+    /// Defaults to only container if there is only one container in the pod.
+    pub container: Option<String>,
+    /// Command to execute.
+    pub command: Option<Vec<String>>,
```

Comment:
It looks like `ExecParams == AttachParams U {command}`

perhaps it's worth factoring out the command and have a common one if you think it generalises?

Reply:
Do you mean to provide a single method to handle both `attach` and `exec` (`attach` if `command` is `None`)?

There shouldn't be any problem doing that other than naming it. If you do `exec` without any `command`, the request is accepted by the server unlike other query parameter errors, but the connection terminates immediately with the following error status:

```json
{
  "apiVersion": "v1",
  "kind": "Status",
  "code": 500,
  "details": {
    "causes": [
      {
        "message": "error executing command in container: failed to exec in container: failed to start exec \"...\": OCI runtime exec failed: exec failed: args must not be empty: unknown"
      }
    ]
  },
  "message": "Internal error occurred: error executing command in container: failed to exec in container: failed to start exec \"...\": OCI runtime exec failed: exec failed: args must not be empty: unknown",
  "metadata": {},
  "reason": "InternalError",
  "status": "Failure"
}
```

```text
Internal error occurred: error executing command in container:
failed to exec in container:
failed to start exec "...":
OCI runtime exec failed:
exec failed:
args must not be empty: unknown
```

Reply:
I was thinking we had could have both `attach` and `exec` use the same `AttachParams`, but have the `exec` method take an extra `command` argument (which based on that error message should probably not be optional).